pymatgen/io/openmm/inputs.py

```diff
@@ -0,0 +1,622 @@
+
+    @staticmethod
+    def _get_coordinates(smiles: Dict[str, int], box: List[float]) -> np.ndarray:
+        """
+        Pack the box with the molecules specified by smiles.
+
+        Args:
+            smiles: keys are smiles and values are number of that molecule to pack
+            box: list of [xlo, ylo, zlo, xhi, yhi, zhi]
+
+        Returns:
+            array of coordinates for each atom in the box.
+        """
+        molecules = []
+        for smile, count in smiles.items():
+            molecules.append(
+                {
+                    "name": smile,
+                    "number": count,
+                    "coords": OpenMMGenerator._smile_to_molecule(smile),
+                }
+            )
+        with tempfile.TemporaryDirectory() as scratch_dir:
+            pw = PackmolBoxGen().get_input_set(molecules=molecules, box=box)
+            pw.write_input(scratch_dir)
+            pw.run(scratch_dir)
+            coordinates = XYZ.from_file(
+                pathlib.Path(scratch_dir, "packmol_out.xyz")
+            ).as_dataframe()
+        raw_coordinates = coordinates.loc[:, "x":"z"].values
+        return raw_coordinates
+
+    @staticmethod
+    def get_box(smiles: Dict[str, int], density: float) -> List[float]:
+        """
+        Calculates the side_length of a cube necessary to contain the given molecules with
+        given density.
+
+        Args:
+            smiles: keys are smiles and values are number of that molecule to pack
+            density: guessed density of the solution, larger densities will lead to smaller cubes.
+
+        Returns:
+            side_length: side length of the returned cube
+        """
+        cm3_to_A3 = 1e24
+        NA = 6.02214e23
+        mols = [OpenMMGenerator._smile_to_molecule(smile) for smile in smiles.keys()]
+        mol_mw = np.array([mol.composition.weight for mol in mols])
+        counts = np.array(list(smiles.values()))
+        total_weight = sum(mol_mw * counts)
+        box_volume = total_weight * cm3_to_A3 / (NA * density)
+        side_length = round(box_volume ** (1 / 3), 2)
+        return [0, 0, 0, side_length, side_length, side_length]
+
+    @staticmethod
+    def _infer_openff_mol(
+        charged_mol: Union[pymatgen.core.Molecule, str, Path]
+    ) -> openff.toolkit.topology.Molecule:
+        if isinstance(charged_mol, (str, Path)):
+            charged_mol = pymatgen.core.Molecule.from_file(str(charged_mol))
+        with tempfile.NamedTemporaryFile() as f:
+            # these next 4 lines are cursed
+            pybel_mol = BabelMolAdaptor(charged_mol).pybel_mol  # pymatgen Molecule
+            pybel_mol.write("mol2", filename=f.name, overwrite=True)  # pybel Molecule
+            rdmol = rdkit.Chem.MolFromMol2File(f.name, removeHs=False)  # rdkit Molecule
+        inferred_mol = openff.toolkit.topology.Molecule.from_rdkit(
+            rdmol, hydrogens_are_explicit=True
+        )  # OpenFF Molecule
```

Comment:
This function uses a `Molecule` from 4 cheminformatics packages in 4 lines of code. It has to be some sort of record.